### ISSUE TITLE
Traefik does not retry failed ACME requests

### ISSUE THREAD

**Issue by larsks** [traefik/traefik#9405]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you do?

I am running Traefik as the ingress provider in a Kubernetes cluster.

If Traefik is unable to reach a given ACME server when it starts up, we see errors such as:
```
traefik-588765f894-7x98v traefik time="2022-09-30T20:47:05Z" level=error msg="Unable to obtain ACME certificate for domains \"postgres.apps.infra.internal\": cannot get ACME client get directory at 'https://step-certificates.step-certificates.svc.cluster.local/acme/acme/directory': acme: error: 0 :: GET :: https://step-certificates.step-certificates.svc.cluster.local/acme/acme/directory :: notFound :: resource not found" providerName=incluster.acme ACME CA="https://step-certificates.step-certificates.svc.cluster.local/acme/acme/directory" rule="Host(`postgres.apps.infra.internal`) && PathPrefix(`/`)" routerName=postgres-pgadmin-postgres-apps-infra-internal@kubernetes
```
The ACME server comes up within a few seconds of these failures, but Traefik does not appear to ever retry the request. It is necessary to delete and re-create Ingress resources in order to force a retry (which succeeds).

That means doing something like:

```
kubectl -n namespace-with-service delete ingress --all
```

Traefik should retry failed requests automatically (with a configurable backoff policy).

### What did you see instead?

Traefik failed to retry the ACME requests, leaving the services in the cluster without valid certificates until we manually intervened.

### What version of Traefik are you using?

Version:      2.9.0-rc5
Codename:     banon
Go version:   go1.19.1
Built:        2022-09-30T13:23:19Z
OS/Arch:      linux/amd64

### What is your environment & configuration?

Kubernetes (KIND) 1.25.0

**Comment by mpl** [comment]
Issue accepted as enhancement, under the new conditions added to the issue description.

**Comment by larsks** [comment]
@mpl I'm moving this (your?) comment from the description to make it more visible. Not sure why this wasn't a comment to begin with:

> **Edit from the maintainers:**
>
> While we acknowledge the problem, and we agree that a possible enhancement would be to retry some steps of the certificate generation process, and only in some cases, we want to make it clear that it is not wise (and therefore not acceptable) to systematically retry everything. Most notably because we should not risk burning through the user's Let's Encrypt quota because of pointless retries.
> Therefore, we'll have to be extra careful in how this is implemented and especially under which conditions, and which steps are retried. For example, a justified retry would be when the lego client fails to be created, then we can retry creating it.

**Comment by larsks** [comment]
Responding to your comment:

I recognize that it's not always wise to retry everything. In this case, I'm operating with a local certificate authority and quotas aren't an issue. This suggests that configurable retries with exponential backoff is probably the right solution -- maybe keep the existing behavior by default, and allow the deployer to configure retries if desired.

**Comment by mpl** [comment]
It was purposefully not a comment. When there's lots of activity on an issue, people (understandably) tend to overlook and not read the comments. So they wouldn't be aware of this additional condition we added.

> @mpl I'm moving this (your?) comment from the description to make it more visible. Not sure why this wasn't a comment to begin with:
> 
> > **Edit from the maintainers:**
> > While we acknowledge the problem, and we agree that a possible enhancement would be to retry some steps of the certificate generation process, and only in some cases, we want to make it clear that it is not wise (and therefore not acceptable) to systematically retry everything. Most notably because we should not risk burning through the user's Let's Encrypt quota because of pointless retries.
> > Therefore, we'll have to be extra careful in how this is implemented and especially under which conditions, and which steps are retried. For example, a justified retry would be when the lego client fails to be created, then we can retry creating it.

**Comment by mpl** [comment]
Sorry, as we think that a vast majority of users are actually Let's Encrypt users, we do not want to add/maintain an option for what we think (for now) is a niche case. Furthermore, we're worried it would be misused by Let's Encrypt users who would shoot themselves in the foot, and who would burn their quota.

Of course, we might reconsider if we get more data points about non-Let's Encrypt users.

**Comment by larsks** [comment]
@mpl isn't this going to be a problem for Lets Encrypt users as well? If Traefik comes up and Lets Encrypt is unavailable (maybe someone misconfigured a security group, or a switch in the data center had a problem, or a firewall rule was misconfigured, etc), this means Traefik won't recover without manual intervention. That seems like an awful failure mode.

I understand your concerns about quotas, but it seems as if at least retrying failed connections would (a) make sense and (b) avoid impacting quotas.

**Comment by mpl** [comment]
@larsks you might be right, but I think our concern was that there's no way to diagnose the cause of a failed connection (quota or not) from where we stand (i.e. users of the lego client).

Since I can't remember for sure at the moment, we should make sure to double-check that, so I'll reopen for now, thanks.

**Comment by axeldunkel** [comment]
I agree not to retry everything. But maybe it would help to retry when the lets encrypt error message contains a timestamp like "retry after 2025-01-05 19:51:49 UTC"?

**Comment by nmengin** [comment]
Hey there,

We think it makes a lot of sense. Unfortunately, this would not make it to our roadmap for a while as we are focused elsewhere.  If you or another community member would like to build it, let us know, and we will work with you to make sure you have all the information needed so that it can be merged.  

We prefer to work with our community members at the beginning of the design process so that we can make sure that we are aligned and can move quickly with the review and merge process.  Let us know here or create a PR before you start, and we will work with you there.  

Don’t forget to check out the [contributor docs](https://github.com/traefik/contributors-guide/blob/master/pr_guidelines.md) and to link the PR to this issue.

**Comment by sami-sweng** [comment]
Hi, on my end I also have issue with no retry but with a tls resolver and a timeout in my case. I'm going to open a separate issue and a PR.

**Comment by Obyka** [comment]
Another use case where retrying after a specific interval could be useful is in a corporate environment with a managed CA. In such environments, certificate requests often require manual approval by a CA manager. After the request is approved, the CSR typically needs to be resent for the certificate to be generated.
